cilium: encryption, subnet IPSec mode configuration may segfault on init

If not all of our preconditions are met, e.g. IP allocation fails or IP lookup fails its possible for enableSubnetIPSec,

```
func (n *linuxNodeHandler) enableSubnetIPsec(v4CIDR, v6CIDR []*net.IPNet)
```

to segfault. For example in an EKS environment where I misconfigured node pools I got this,

```
level=debug msg="IPsec enable succeeded" local-ip=0.0.0.0/0 reason="CNI Out IPv4" remote-ip=192.168.0.0/16 spi=3 subsys=linux-datapath
level=error msg="IPsec enable failed" error="Link not found" local-ip=0.0.0.0/0 reason="getV4LinkLocalIP failed" remote-ip=192.168.0.0/16 spi=3 subsys=linux-datapath
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2009397]

goroutine 1 [running]:
github.com/cilium/cilium/pkg/datapath/linux/ipsec.UpsertIPsecEndpoint(0x0, 0xc000f9a0c0, 0xc000c8bb90, 0x299e187, 0x8, 0xc000c8bb01, 0x275f903, 0x0, 0x0)
        /go/src/github.com/cilium/cilium/pkg/datapath/linux/ipsec/ipsec_linux.go:334 +0x37
github.com/cilium/cilium/pkg/datapath/linux.(*linuxNodeHandler).enableSubnetIPsec(0xc0009ae000, 0xc000baada0, 0x2, 0x2, 0x46308e8, 0x0, 0x0)
        /go/src/github.com/cilium/cilium/pkg/datapath/linux/node.go:534 +0x2d4
github.com/cilium/cilium/pkg/datapath/linux.(*linuxNodeHandler).nodeUpdate(0xc0009ae000, 0x0, 0xc000c5a540, 0x1, 0x2, 0xc000baada0)
        /go/src/github.com/cilium/cilium/pkg/datapath/linux/node.go:922 +0x21b
github.com/cilium/cilium/pkg/datapath/linux.(*linuxNodeHandler).NodeConfigurationChanged(0xc0009ae000, 0x2329, 0x22aa, 0x22dc, 0x22f7, 0x100, 0x46308e8, 0x0, 0x0, 0x1000000000001, ...)
        /go/src/github.com/cilium/cilium/pkg/datapath/linux/node.go:1398 +0x487
```

The system wasn't going to come up anyways so it would have failed and crash looped regardless, but still issuing a splat because we don't handle errors nicely is not user friendly. Offending code here,

```
                        linkAddr, err := getV4LinkLocalIp()
                        if err != nil {
                                upsertIPsecLog(err, "getV4LinkLocalIP failed", ipsecIPv4Wildcard, cidr, spi)
                        }
                        spi, err := ipsec.UpsertIPsecEndpoint(linkAddr, ipsecIPv4Wildcard, cidr, ipsec.IPSecDirIn, zeroMark)
                        upsertIPsecLog(err, "CNI In IPv4", linkAddr, ipsecIPv4Wildcard, spi)
```
Notice we just ignore the error and try to upsert anyways with linkAddr == null. Clearly not going to work. Above was caused by user error but like noted above we should be a bit friendlier with our errors.

--- a/daemon/cmd/daemon_main.go
+++ b/daemon/cmd/daemon_main.go
@@ -1530,6 +1530,10 @@ func initEnv(vp *viper.Viper) {
 		)
 	}
 
+	if option.Config.IPAM == ipamOption.IPAMENI && option.Config.TunnelingEnabled() {
+		log.Fatalf("Cannot specify IPAM mode %s in tunnel mode.", option.Config.IPAM)
+	}
+
 	if option.Config.IPAM == ipamOption.IPAMMultiPool {
 		if option.Config.TunnelingEnabled() {
 			log.Fatalf("Cannot specify IPAM mode %s in tunnel mode.", option.Config.IPAM)

--- a/pkg/datapath/linux/ipsec.go
+++ b/pkg/datapath/linux/ipsec.go
@@ -55,6 +55,9 @@ func (n *linuxNodeHandler) getLinkLocalIP(family int) (net.IP, error) {
 	if err != nil {
 		return nil, err
 	}
+	if len(addr) == 0 {
+		return nil, fmt.Errorf("error retrieving link local IP (family %d): no addresses found", family)
+	}
 	return addr[0].IPNet.IP, nil
 }
 

